### PR TITLE
Assignment2 Rahat Bhatia and Paul Aguilar

### PR DESCRIPTION
**Pull request by rahat15**
@rahat15 @pja6

### CODE REVIEW

File: Logger/BaseLoggerMixins.cs [hunk 1]
```diff
@@ -1,6 +1,54 @@
...
+{
+    public static void Error(this BaseLogger? logger, string message, params object[] args)
+    {
+        if (logger == null)
+        {
+            throw new System.ArgumentNullException(nameof(logger));
+        }
+        else
+        {
+            logger.Log(LogLevel.Error, string.Format(CultureInfo.CurrentCulture, message, args));
```
Comment (SaaWest) [inline Logger/BaseLoggerMixins.cs:15]: I think the following could be different
```suggestion
            logger.Log(LogLevel.Error, string.Format(CultureInfo.CurrentCulture, message, args));
```
to
```suggestion
            string formatMess = string.Format(message, args);
            logger.Log(LogLevel.Error, formatMess);

```

**Review by SaaWest — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:

    The current date/time ✔
    The name of the class that created the logger ✔
    The log level ✔
    The message ✔
    The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"

The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for

    Error, ✔
    Warning, ✔
    Information, and ✔
    Debug. ✔
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

EXTRA CREDIT

    Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ❌
    Implement the factory pattern using static abstract methods on a Logger interface instead. ❌

File: .idea/.idea.Logger/.idea/.name
Comment (BenjaminMichaelis) [inline .idea/.idea.Logger/.idea/.name:1]: These files should be excluded from the repo by adding to the gitignore

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 2]
```diff
@@ -14,24 +14,108 @@ public void Error_WithNullLogger_ThrowsException()
...
+
+
     [TestMethod]
     public void Error_WithData_LogsMessage()
     {
         // Arrange
         var logger = new TestLogger();
 
         // Act
-        //logger.Error("Message {0}", 42);
```
Comment (BenjaminMichaelis) [inline Logger.Tests/BaseLoggerMixinsTests.cs:29]: By making this change, you are no logger checking that that replacement is happening

File: Logger.Tests/FileLoggerTests.cs [hunk 3]
```diff
@@ -1,9 +1,111 @@
...
+
+using NUnit.Framework.Internal;
+
+namespace Logger.Tests;
+
+[TestClass]
+public class FileLoggerTests
+{
+    //private FileLogger? _fileLogger;
+    private readonly string _filePath = "text.txt";
```
Comment (BenjaminMichaelis) [inline Logger.Tests/FileLoggerTests.cs:16]: Use a property here instead of a field

File: Logger.Tests/FileLoggerTests.cs [hunk 4]
```diff
@@ -1,9 +1,111 @@
...
-
-}
+using System;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+using NUnit.Framework.Internal;
```
Comment (BenjaminMichaelis) [inline Logger.Tests/FileLoggerTests.cs:8]: What is this used for?

File: Logger.Tests/LogFactoryTests.cs [hunk 5]
```diff
@@ -1,9 +1,33 @@
...
+
+[TestClass]
+public class LogFactoryTests
+{
+    [TestMethod]
+    public void CreateLogger_CheckClassName_Success()
+    {
+
+        var logFactory = new LogFactory();
+        string filePath = "C:\\path\\to\\your\\test.txt";
```
Comment (BenjaminMichaelis) [inline Logger.Tests/LogFactoryTests.cs:15]: This won't work on a linux machine. I would use a relative path. Also use something like `Path.Combine` and `Path.Join`.

File: Logger.Tests/LogFactoryTests.cs [hunk 6]
```diff
@@ -1,9 +1,33 @@
...
+        logFactory.ConfigureFileLogger(filePath);
+
+        var fileLogger = logFactory.CreateLogger("FileLogger");
+
+        Assert.IsNotNull(fileLogger, "FileLogger should not be null");
+        Assert.AreEqual("FileLogger", fileLogger?.ClassName);
```
Comment (BenjaminMichaelis) [inline Logger.Tests/LogFactoryTests.cs:21]: Why not use nameof here?

File: Logger/BaseLogger.cs [hunk 7]
```diff
@@ -3,5 +3,8 @@
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+
+    public string? ClassName { get; set; }
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:7]: Why is this nullable?

File: Logger/BaseLoggerMixins.cs [hunk 8]
```diff
@@ -1,6 +1,57 @@
...
-{
-
-}
+﻿using System.Globalization;
+
+namespace Logger;
+
+public static class BaseLoggerMixins
+{
+    public static void Error(this BaseLogger? logger, string message, params object[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:7]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger/FileLogger.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    public class FileLogger : BaseLogger
+    {
+   
+
+        public string FilePath
+        {
+            get;
+            set;
+        }
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:16]: This should be one line

File: Logger/FileLogger.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        public FileLogger(string filePath)
+        {
+
+            this.FilePath = filePath;
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:20]: `this` is not needed here

File: Logger/LogFactory.cs [hunk 11]
```diff
@@ -1,10 +1,41 @@
...
+
+    public string FilePath
+    {
+        get => _filePath!; //telling compiler it's okay to be null in this case
+        set
+        {
+            //if value is null throw exception
+            _filePath = value ?? throw new ArgumentNullException(nameof(value));
+        }
+    }
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:22]: Nice job with this pattern!

File: Logger/LogFactory.cs [hunk 12]
```diff
@@ -1,10 +1,41 @@
...
+
+    public BaseLogger? CreateLogger(string className)
+    {
+        ConfigureFileLogger(FilePath);
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:26]: Why call this method from here? What is this doing?

**Pull request closed without merging**